sharedbuffers/mapped_struct.py

```diff
@@ -342,6 +342,194 @@ def unpack_from(cls, buf, offs, idmap = None):
         key, values = mapped_list.unpack_from(buf, offs, idmap)
         return cls(zip(key, values))
 
+
+def _upsize(x):
+    "Round up X to the next power of 2"
+    for p2 in (1, 2, 4, 8, 16, 32, 64):
+        x |= x >> p2
+    return x + 1
+
+def _add_kv(indexes, kvpairs, key, val):
+    l2 = len(indexes) / 2
+    hval = hash(key)
+    index = hval % l2
+    nprobe = 1
+
+    while True:
+        i2 = index + index
+        if indexes[i2] == -1:
+            indexes[i2] = len(kvpairs)
+            indexes[i2 + 1] = hval
+            kvpairs.extend((key, val))
+            return
+
+        index = (index + nprobe) % l2
+        nprobe += 1
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(
+        index_table=tuple,
+        kvpairs=proxied_list,
+        mask=int,
+        nelems=int)
+
+    def __init__(self, mask, nelems, index_table, kvpairs):
+        self.mask = mask
+        self.nelems = nelems
+        self.index_table = index_table
+        self.kvpairs = kvpairs
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        """
+        The binary format of a proxied dictionary is as follows:
+        | size           | type            | attribute name
+        |  8             | long            | mask
+        |  8             | long            | nelems
+        | $mask + 1 * 16 | array of long's | index_table
+        | $nelems        | proxied_list    | kvpairs
+
+        In other words, there are 2 tables: an index vector and a key/value
+        array. The index vector contains the positions in the second array,
+        and their hash values, to speed up comparisons. Unused buckets have
+        a position of -1. The key/value array is packed into a proxied list.
+        """
+
+        xlen = len(obj)
+        size = _upsize(xlen + xlen / 2)   # Aim for a load factor of ~50%
+        indexes = [-1, 0] * size
+        kvpairs = []
+        nelem = 0
+
+        for key, val in obj.iteritems():
+            _add_kv(indexes, kvpairs, key, val)
+            nelem += 1
+
+        len_m1 = len(indexes) / 2 - 1
+        struct.pack_into("=QQ", buf, offs, len_m1, nelem)
+        offs += 16
+        abuf = buffer(array.array('l', indexes))
+        buf[offs:offs + len(abuf)] = abuf
+        offs += len(abuf)
+
+        return offs + mapped_tuple.pack_into(kvpairs, buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        mask, nelems = struct.unpack_from("=QQ", buf, offs)
+        offs += 16
+        idx_size = (mask + 1) * 16
+        index_table = array.array('l')
+        index_table.fromstring(buffer(buf[offs:offs + idx_size]))
+        index_table = tuple(index_table)
```

Comment:
This is not a proxied dict. In here you're materializing the whole dict in memory. I very much prefer mapped_dict's implementation in that case, so thanks but no thanks.

The whole point of proxies is to avoid copying the whole data structure into process private memory (ie: copying). The `unpack_from` operation must be O(1), it must construct a class that can then look up the required item in less than O(N).

Reply:
Would it be enough to change the index table into a numpy array and then use that proxy ?

Reply:
No. The current lookup functions are broken when rehashing nontrivial hash maps with collisions (the probe sequence isn't preserved crossing the 2^n boundary). You need a correct implementation, and hash tables are tricky and very unfriendly for disk access.

Sharedbuffers are mapped from disk and they can be bigger than RAM, so we have an algorithm already that is optimized for disk friendliness, compactness and is less than O(N), which is the one the various IdMappers use.

Use the id mappers. If not the classes at least the algorithms and helpers.